test/relayclient/RelayProvider.test.ts

```diff
@@ -0,0 +1,274 @@
+    // @ts-ignore
+    TestRecipient.web3.setProvider(relayProvider)
+    const res = await testRecipient.emitMessage('hello world', {
+      from: gasLess,
+      forceGasPrice: '0x51f4d5c00',
+      // TODO: for some reason estimated values are crazy high!
+      gas: '100000',
+      paymaster,
+      forwarder
+    })
+    assert.equal(res.logs[0].event, 'SampleRecipientEmitted')
+    assert.equal(res.logs[0].args.message, 'hello world')
+    assert.equal(res.logs[0].args.realSender, gasLess)
+  })
+
+  describe('_ethSendTransaction', function () {
+    const id = 777
+    let testRecipient: TestRecipientInstance
+    let gsnConfig: GSNConfig
+    let jsonRpcPayload: JsonRpcPayload
+    let dependencyTree: TmpDependencyTree
+
+    before(async function () {
+      const TestRecipient = artifacts.require('TestRecipient')
+      testRecipient = await TestRecipient.new()
+      const forwarder = await testRecipient.getTrustedForwarder()
+      gsnConfig = configureGSN({ relayHubAddress: relayHub.address })
+      dependencyTree = RelayClient.getDefaultDependencies(underlyingProvider, gsnConfig)
+      // call to emitMessage('hello world')
+      jsonRpcPayload = {
+        jsonrpc: '2.0',
+        id,
+        method: 'eth_sendTransaction',
+        params: [
+          {
+            from: gasLess,
+            gas: '0x186a0',
+            gasPrice: '0x4a817c800',
+            forceGasPrice: '0x51f4d5c00',
+            paymaster,
+            forwarder,
+            to: testRecipient.address,
+            data: '0x2ac0df260000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000b68656c6c6f20776f726c64000000000000000000000000000000000000000000'
+          }
+        ]
+      }
+    })
+
+    it('should call callback with error if relayTransaction throws', function () {
+      const badRelayClient = new BadRelayClient(true, false, dependencyTree, gsnConfig.relayHubAddress, gsnConfig.relayClientConfig)
+      const relayProvider = new RelayProvider(badRelayClient, underlyingProvider, gsnConfig)
+      const promisified = new Promise((resolve, reject) => relayProvider._ethSendTransaction(jsonRpcPayload, (error: Error | null): void => {
+        reject(error)
+      }))
+      return expect(promisified).to.be.eventually.rejectedWith(`Rejected relayTransaction call - should not happen. Reason: Error: ${BadRelayClient.message}`)
+    })
+
+    it('should call callback with error containing relaying results dump if relayTransaction does not return a transaction object', function () {
+      const badRelayClient = new BadRelayClient(false, true, dependencyTree, gsnConfig.relayHubAddress, gsnConfig.relayClientConfig)
+      const relayProvider = new RelayProvider(badRelayClient, underlyingProvider, gsnConfig)
+      const promisified = new Promise((resolve, reject) => relayProvider._ethSendTransaction(jsonRpcPayload, (error: Error | null): void => {
```

Comment:
better use `promisify`: it handles both rejects and resolved value.

Reply:
We don't use `promisify` anywhere in code currently. I don't need it for this simple thing.

Reply:
(except that you do it 4 times in this file, and maybe elsewhere. the syntax of promisify is much simpler (hence much less possible mistakes, and less eye-strain)
(not critical, tough)